### PR TITLE
fix clippy warnings

### PR DESCRIPTION
**Pull request by LizardWizzard**
I've fixed all clippy warnings for our code base according to clippy's default configuration. Most of the changes are pretty mechanical. If something doesnt look right to you lets figure out which lints we should ignore and where to put #[allow] marks :)

This is a starting point to running clippy in the CI and adding it as a pre commit check similar to rustfmt

Command to run clippy: `cargo clippy --all-targets --all-features --all --tests -- -D warnings`

### CODE REVIEW

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -1353,12 +1353,10 @@ impl LayeredTimeline {
                 } else {
                     result.ondisk_nonrelfiles_dropped += 1;
                 }
+            } else if doomed_layer.get_seg_tag().rel.is_relation() {
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:1356]: I don't think this is an improvement, I'd like to keep the if-branches symmetrical.

Maybe there's a better way to refactor this though. This is a recurring pattern in the function, checking `rel.is_relation()` and incrementing either `relfiles_something` or `nonrelfiles_something`.

File: pageserver/src/layered_repository/layer_map.rs [hunk 2]
```diff
@@ -122,11 +122,7 @@ impl LayerMap {
     pub fn get_open(&self, tag: &SegmentTag) -> Option<Arc<InMemoryLayer>> {
         let segentry = self.segs.get(tag)?;
 
-        if let Some(open) = &segentry.open {
-            Some(Arc::clone(open))
-        } else {
-            None
-        }
+        segentry.open.as_ref().map(|open| Arc::clone(open))
```
Comment (hlinnaka) [inline pageserver/src/layered_repository/layer_map.rs:125]: I think the `if let` construct is more readable. At least to my untrained eye, the map() syntax takes a while to parse. It makes me think there's iteration happening but it's just an Option rather than a Collection. And the `as_ref()` adds noise.

File: pageserver/src/layered_repository/layer_map.rs [hunk 3]
```diff
@@ -292,11 +288,9 @@ impl LayerMap {
 
     /// Return the oldest in-memory layer.
     pub fn peek_oldest_open(&self) -> Option<Arc<InMemoryLayer>> {
-        if let Some(opensegentry) = self.open_segs.peek() {
-            Some(Arc::clone(&opensegentry.layer))
-        } else {
-            None
-        }
+        self.open_segs
```
Comment (hlinnaka) [inline pageserver/src/layered_repository/layer_map.rs:291]: same here, the map() doesn't look as readable to me.

File: pageserver/src/restore_local_repo.rs [hunk 4]
```diff
@@ -26,7 +26,7 @@ use zenith_utils::lsn::Lsn;
 const MAX_MBR_BLKNO: u32 =
     pg_constants::MAX_MULTIXACT_ID / pg_constants::MULTIXACT_MEMBERS_PER_PAGE as u32;
 
-const ZERO_PAGE: Bytes = Bytes::from_static(&[0u8; 8192]);
+const ZERO_PAGE: &[u8; 8192] = &[0u8; 8192];
```
Comment (hlinnaka) [inline pageserver/src/restore_local_repo.rs:29]: I'm surprised by this change. It makes all the places that use ZERO_PAGE more verbose. What was the warning?

File: pageserver/src/restore_local_repo.rs [hunk 5]
```diff
@@ -483,9 +483,9 @@ pub fn save_decoded_record(
                 },
                 rpageno,
                 lsn,
-                ZERO_PAGE,
+                Bytes::from_static(ZERO_PAGE),
             )?;
-        } else if info == pg_constants::XLOG_MULTIXACT_ZERO_OFF_PAGE {
+        } else if info == pg_constants::XLOG_MULTIXACT_ZERO_MEM_PAGE {
```
Comment (hlinnaka) [inline pageserver/src/restore_local_repo.rs:488]: Now that's fixing an actual bug, not just a clippy warning :-). https://github.com/zenithdb/zenith/issues/453

Let's do that as a separate PR.

File: pageserver/src/restore_local_repo.rs [hunk 6]
```diff
@@ -744,14 +741,14 @@ fn transaction_id_precedes(id1: TransactionId, id2: TransactionId) -> bool {
      * comparison.  If both are normal, do a modulo-2^32 comparison.
      */
 
-    if !(id1 >= pg_constants::FIRST_NORMAL_TRANSACTION_ID)
-        || !(id2 >= pg_constants::FIRST_NORMAL_TRANSACTION_ID)
+    if (id1 < pg_constants::FIRST_NORMAL_TRANSACTION_ID)
```
Comment (hlinnaka) [inline pageserver/src/restore_local_repo.rs:744]: The reason this was written this way was to match the code in Postgres TransactionIdPrecedes. But I agree it was confusing. I'm OK with changing this the way proposed here, but alternatively we could introduce a `TransactionIdIsNormal` function, like in Postgres, and use that here. Then this function would match the Postgres function more closely, and it would be easier to see that it's equivalent.

File: walkeeper/src/safekeeper.rs [hunk 7]
```diff
@@ -378,6 +381,7 @@ where
     }
 
     /// Handle request to append WAL.
+    #[allow(clippy::comparison_chain)]
```
Comment (hlinnaka) [inline walkeeper/src/safekeeper.rs:384]: What did clippy suggest here instead? Is it ugly?

File: zenith_utils/src/pq_proto.rs [hunk 8]
```diff
@@ -164,7 +164,7 @@ impl FeStartupMessage {
             Err(e) => return Err(e.into()),
         };
 
-        if len < 4 || len > MAX_STARTUP_PACKET_LENGTH {
+        if !(4..=MAX_STARTUP_PACKET_LENGTH).contains(&len) {
```
Comment (hlinnaka) [inline zenith_utils/src/pq_proto.rs:167]: This is much less readable than the original IMHO

**Review by hlinnaka — COMMENTED** [review]
Nice cleanups overall.

File: pageserver/src/restore_local_repo.rs (hunk 4 above)
Comment (LizardWizzard) [reply pageserver/src/restore_local_repo.rs]: Warning was about having const of type which is interior mutable, I've revisited it, initially I thought that replacing const with static would be enough, but I was concerned about possibility of mutation which is impossible because Bytes are immutable (there is a differrent type, BytesMut which can be mutable), so this changes to be just static, and ZERO_PAGE.clone() when using it. Also I found that we already use `static ZERO_PAGE: Bytes` probably we should create some zenith_consts to share that

File: walkeeper/src/safekeeper.rs (hunk 7 above)
Comment (LizardWizzard) [reply walkeeper/src/safekeeper.rs:384]: here is the reference with an example https://rust-lang.github.io/rust-clippy/master/index.html#comparison_chain

TLDR, instead of this
```rust
if x > y {
    a()
} else if x < y {
    b()
}
```
use that
```rust
 match x.cmp(&y) {
     Ordering::Greater => a(),
     Ordering::Less => b(),
     Ordering::Equal => {}
 }
```
what do you think about cmp/match option?

File: pageserver/src/restore_local_repo.rs (hunk 6 above)
Comment (LizardWizzard) [reply pageserver/src/restore_local_repo.rs:744]: Didn't know that, I like your idea about separate function, will do that. Will posgres_ffi be a suitable place for that?

File: pageserver/src/layered_repository.rs (hunk 1 above)
Comment (LizardWizzard) [reply pageserver/src/layered_repository.rs:1356]: Doesn't look better to me neither, I've spent a few minutes on that, and decided to apply the suggestion. I see the following option:
```rust
match (doomed_layer.is_dropped(), doomed_layer.get_seg_tag().rel.is_relation()) {
    (true, true) => result.ondisk_relfiles_dropped += 1,
    (true, false) => result.ondisk_nonrelfiles_dropped += 1,
    (false, true) => result.ondisk_relfiles_removed += 1,
    (false, false) => result.ondisk_nonrelfiles_removed += 1,
}
```
what do you think about it?

Comment (hlinnaka) [reply pageserver/src/layered_repository.rs:1356]: Works for me

File: pageserver/src/restore_local_repo.rs (hunk 6 above)
Comment (hlinnaka) [reply pageserver/src/restore_local_repo.rs:744]: Yeah. Should probably move `transaction_id_precedes` there, too.

File: pageserver/src/restore_local_repo.rs (hunk 5 above)
Comment (LizardWizzard) [reply pageserver/src/restore_local_repo.rs:488]: Done. See #499

File: zenith_utils/src/pq_proto.rs (hunk 8 above)
Comment (LizardWizzard) [reply zenith_utils/src/pq_proto.rs:167]: What do you think is better, put an allow mark here, or ignore this lint globally?

File: pageserver/src/layered_repository/layer_map.rs (hunk 2 above)
Comment (funbringer) [reply pageserver/src/layered_repository/layer_map.rs:125]: > It makes me think there's iteration happening

Your intuition is correct: iteration is exactly what's happening here. We're iterating over a collection comprised of _no more than_ one element. Unfortunately, Rust is unable to define a common trait for mappable types (Option, Result, Vec, HashMap, the list goes on), because its type system has no notion of higher-kinded types.

Personally, I'd go as far as to write the following:

```rust
segentry.open.as_ref().map(Arc::clone)
```

> And the as_ref() adds noise

This is subjective. As a matter of fact, `as_ref` allows one to chain calls when a subsequent method requires a reference to the data.

File: pageserver/src/restore_local_repo.rs (hunk 6 above)
Comment (LizardWizzard) [reply pageserver/src/restore_local_repo.rs]: done in #501

File: pageserver/src/layered_repository/layer_map.rs (hunk 2 above)
Comment (LizardWizzard) [reply pageserver/src/layered_repository/layer_map.rs:125]: I'm personally leaning towards @funbringer's suggestion, for me map looks pretty organic here

Comment (hlinnaka) [reply pageserver/src/layered_repository/layer_map.rs:125]: Ok, I'm outvoted by you, @funbringer and Clippy then :-).

File: zenith_utils/src/pq_proto.rs (hunk 8 above)
Comment (LizardWizzard) [reply zenith_utils/src/pq_proto.rs:167]: put an allow in `zenith_utils/lib.rs`

File: pageserver/src/layered_repository/layer_map.rs (hunk 3 above)
Comment (LizardWizzard) [reply pageserver/src/layered_repository/layer_map.rs]: This change become irrelevant with rebase to current main

**Comment by LizardWizzard** [comment]
There is one warning left. 
```
error: large size difference between variants
   --> zenith_utils/src/sock_split.rs:112:5
    |
112 | /     Tls {
113 | |         stream: BufStream,
114 | |         session: rustls::ServerSession,
115 | |     },
    | |_____^ this variant is 568 bytes
    |
    = note: `-D clippy::large-enum-variant` implied by `-D warnings`
note: and the second-largest variant is 40 bytes:
   --> zenith_utils/src/sock_split.rs:111:5
    |
111 |     Tcp(BufStream),
    |     ^^^^^^^^^^^^^^
help: consider boxing the large fields to reduce the total size of the enum
   --> zenith_utils/src/sock_split.rs:112:5
    |
112 | /     Tls {
113 | |         stream: BufStream,
114 | |         session: rustls::ServerSession,
115 | |     },
    | |_____^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant

error: aborting due to previous error
```
@patins what do you think about boxing in this case?

I think this can be merged in current state to stop the rebases and I'll submit another PR with CI handling
